Generalized Hyperbolic Distribution (#13298)
scipy/scipy: doc/source/tutorial/stats/continuous_genhyperbolic.rst

The "hyperbolic" characterization refers to the fact that shape the log-probability distribution can be described as an hyperbola. Hyperbolic distributions are sometime referred as semi-fat tail because their probability density decrease slower than "sub-hyperbolic" distributions (e.g. normal distribution, whose log-probability decreases quadratically), but faster than other "extreme value" distributions (e.g. `pareto` distribution, whose log-probability decreases logarithmically).

Functions
---------

Different parameterization exists in literature, scipy implementing the location-scale as for Prause (1999)

.. math::

   \begin{eqnarray*} f(x, p, \hat{\alpha}, \hat{\beta}, \text{location}, \text{scale}) &=&
            \frac{(\hat{\alpha}^2 - \hat{\beta}^2)^{\lambda/2}}
            {\sqrt{2\pi}\hat{\alpha}^{\lambda-0.5}
            K_\lambda\Big(\sqrt{\hat{\alpha}^2 - \hat{\beta}^2}\Big)}
            e^{\hat{\beta}(x)} \times \frac{K_{\lambda - 1/2}
            (\hat{\alpha} \sqrt{1 + x^2})}
            {(\sqrt{1 + x^2})^{1/2 - \lambda}} \end{eqnarray*}

for:

- :math:`x, \lambda \in \mathbb{R}`

> - :math:`|\hat{\beta}| \lt \hat{\alpha}` if :math:`p \ge 0`

Reviewer: Did you intend to use `p` instead of `lambda` above?

Author: probably yes, i will address this together with https://github.com/scipy/scipy/pull/13298#discussion_r556778096

Reviewer: OK, just ping me when this is ready to review again!

- :math:`|\hat{\beta}| \lt \hat{\alpha}` if :math:`p = 0`
- :math:`|\hat{\beta}| \leq \hat{\alpha}` if :math:`p \lt 0`
- :math:`K_{\lambda}(.)` denotes the modified Bessel function of the second kind and order :math:`p` (`scipy.special.kn`)

which derives from the original parameterization in  Barndorff (1978) by setting:

- :math:`p = \lambda`
- :math:`\hat{\alpha} = \delta \alpha`
- :math:`\hat{\beta} = \delta \beta`
- :math:`\text{location} = \mu`
- :math:`\text{scale} = \delta`

> Random variates for the `scipy.stats.genhyperbolic` can be efficiently sampled from the above-mentioned normal variance-mean mixture where `scipy.stats.geninversegauss` is parametrized as :math:`GIG = f(p = p, b = \Big(\hat{\alpha}^2 - \hat{\beta}^2\Big)^{1/2}, \text{loc} = \text{location}, \text{scale} = \Big(\hat{\alpha}^2 - \hat{\beta}^2\Big)^{ - 1/2})` so that :math:`GH(p, \hat{\alpha}, \hat{\beta}) = \hat{\beta} GIG + \sqrt{GIG} N(0,1)`

Reviewer: ```suggestion
Random variates for the `scipy.stats.genhyperbolic` can be efficiently sampled from the above-mentioned normal variance-mean mixture where `scipy.stats.geninvgauss` is parametrized as :math:`GIG = f(p = p, b = \Big(\hat{\alpha}^2 - \hat{\beta}^2\Big)^{1/2}, \text{loc} = \text{location}, \text{scale} = \Big(\hat{\alpha}^2 - \hat{\beta}^2\Big)^{ - 1/2})` so that :math:`GH(p, \hat{\alpha}, \hat{\beta}) = \hat{\beta} GIG + \sqrt{GIG} N(0,1)`
```

.. _continuous-genhyperbolic:

Generalized Hyperbolic Distribution
===================================

> The Generalized Hyperbolic Distribution is defined as the normal variance-mean mixture with Generalized Inverse Gaussian distribution as mixing distribution.

Author: This should be pretty aligned with what wiki says

Functions
---------

Different parameterization exists in literature, scipy implementing the location-scale as for Prause (1999)

.. math::
   :nowrap:

    \begin{eqnarray*}
        f(x, p, a, b) & = &
        \frac{(a^2 - b^2)^{p/2}}
        {\sqrt{2\pi}a^{p-0.5}
        K_p\Big(\sqrt{a^2 - b^2}\Big)}
        e^{b(x)} \times \frac{K_{p - 1/2}
        (a \sqrt{1 + x^2})}
        {(\sqrt{1 + x^2})^{1/2 - p}}
    \end{eqnarray*}

for:

-  :math:`x, p \in ( - \infty; \infty)`
-  :math:`|b| < a` if :math:`p \ge 0`
-  :math:`|b| \le a` if :math:`p < 0`
-  :math:`K_{p}(.)` denotes the modified Bessel function of the second kind and order :math:`p` (`scipy.special.kn`)

which derives from the original :math:`(\lambda, \alpha, \beta, \delta, \mu)` parameterization in  Barndorff (1978) by setting:

-  :math:`\lambda = p`
-  :math:`\alpha = \frac{a}{\delta} = \frac{\hat{\alpha}}{\delta}`
-  :math:`\beta = \frac{b}{\delta} = \frac{\hat{\beta}}{\delta}`
-  :math:`\delta = \text{scale}`

> -  :math:`\mu = \text{location}`

Author: ok will explicit that

The "generalized" characterization suggests the fact that this distribution is a superclass of several other probability distribution, for instance:

-  :math:`f(p = \nu,  a = 0, b = 0, \text{loc} = 0, \text{scale} = \sqrt{\delta})` has a Student's t-distribution (`scipy.stats.t`) with :math:`\nu` degrees of freedom.
-  :math:`f(p = 1, a = \hat{\alpha}, b = \hat{\beta}, \text{loc} = \mu, \text{scale} = \delta)` has a Hyperbolic Distribution.
-  :math:`f(p = - 1/2, a = \hat{\alpha}, b = \hat{\beta}, \text{loc} = \mu, \text{scale} = \delta)` has a Normal Inverse Gaussian Distribution (`scipy.stats.norminvgauss`).
-  :math:`f(p = 1, a = 0, b = 0, loc = \mu, \text{scale} = \delta)` has a Laplace Distribution (`scipy.stats.laplace`) for :math:`\delta \rightarrow 0`

Examples
--------

> It is useful to understand how the parameters affect the shape of the distribution. While it is fairly straightforward to interpret the meaning of :math:`b` as the skeweness, understanding the difference among :math:`a` and :math:`p` is not as obvious, as both parameters do in fact affect the kurtosis of the distribution. :math:`a` can be interpreted as the speed of the decay of the probability density (where :math:`a > 1` the asymptotic decay is faster than :math:`log_e` and vice versa) or - equivalently - as the slope of the log-probability hyperbola asymptote (where :math:`a > 1` decay is faster than :math:`|1|` and vice versa), while :math:`p` can be se as the width of the shoulders of the probability density distribution (where :math:`p < 1` narrow shoulders and vice versa) or - equivalently - as the shape of the log-probability hyperbola, which is convex :math:`p < 1` and concave otherwise.

Author: Not really i am afraid.  Here i refer to the peakedness of the pdf . "Narrow" could _probably_ be described in terms of [higher moments](https://en.wikipedia.org/wiki/Moment_(mathematics)#Higher_moments) https://en.wikipedia.org/wiki/Moment_(mathematics)#Higher_moments but to be fair the whole text should be read in conjunction with the two plots generated by the code below (i know they are not generated) so the narrow shoulders can be visualized

.. _continuous-genhyperbolic:

Generalized Hyperbolic Distribution
===================================

The Generalized Hyperbolic Distribution is defined as the normal variance-mean mixture with Generalized Inverse Gaussian distribution as the mixing distribution.
The "hyperbolic" characterization refers to the fact that the shape of the log-probability distribution can be described as a hyperbola. Hyperbolic distributions are sometime referred to as semi-fat tailed because their probability density decrease slower than "sub-hyperbolic" distributions (e.g. normal distribution, whose log-probability decreases quadratically), but faster than other "extreme value" distributions (e.g. `pareto` distribution, whose log-probability decreases logarithmically).

Functions
---------

> Different parameterization exists in the literature; SciPy implements the location-scale as in Prause (1999).

Reviewer: ```suggestion
Different parameterizations exist in the literature; SciPy implements the "4th parametrization" in Prause (1999).
```
I couldn't find "location-scale" in the Prause reference cited below, but this looks like the "4th parametrization" in that paper.

Note that year 1999 isn't mentioned in the references below, so maybe there was a different paper? But we should use the references we cite and vice-versa

Author: You are right [this](https://freidok.uni-freiburg.de/fedora/objects/freidok:15/datastreams/FILE1/content) is the phd dissertation (not a paper, at least not that i could find) from which the 2002 paper is likely taken, I will reference it here and in the docstring.

.. math::
   :nowrap:

>     \begin{eqnarray*}

Reviewer: For the record, I copied down this PDF, did the substitutions, and got the original parameterization as written in  Barndorff .

        f(x, p, a, b) & = &
        \frac{(a^2 - b^2)^{p/2}}
        {\sqrt{2\pi}a^{p-0.5}
        K_p\Big(\sqrt{a^2 - b^2}\Big)}

>         e^{b(x)} \times \frac{K_{p - 1/2}

Reviewer: ```suggestion
        e^{bx} \times \frac{K_{p - 1/2}
```
Perfectly correct as is, but I read it as a function instead of multiplication. Looks like a holdout from when the location was in the parentheses, too?

Author: yes indeed

        (a \sqrt{1 + x^2})}
        {(\sqrt{1 + x^2})^{1/2 - p}}
    \end{eqnarray*}

> The probability density above is defined in the “standardized” form. To shift and/or scale the distribution use the loc and scale parameters. Specifically, :math:`f(x, p, a, b, \text{loc}, \text{scale})` is identically equivalent to :math:`\frac{1}{\text{scale}}f(y, p, a, b)` with :math:`y = \frac{1}{\text{scale}}(x - \text{loc})`.
> 

Reviewer: ```suggestion
```

> for:

Reviewer: Thanks for adding this, but I think it was inserted in the middle of a sentence.

Author: How about before the `pdf` equation ?

Reviewer: Since `loc` and `scale` aren't mentioned until the Barndorff parameterization, I would put it right above that. I'll suggest it.

-  :math:`x, p \in ( - \infty; \infty)`
-  :math:`|b| < a` if :math:`p \ge 0`
-  :math:`|b| \le a` if :math:`p < 0`
-  :math:`K_{p}(.)` denotes the modified Bessel function of the second kind and order :math:`p` (`scipy.special.kn`)

> which derives from the original :math:`(\lambda, \alpha, \beta, \delta, \mu)` parameterization in  Barndorff (1978) by setting:

Reviewer: ```suggestion
The probability density above is defined in the "standardized" form. To shift and/or scale the distribution use the :math:`\text{loc}` and :math:`\text{scale}` parameters. Specifically, :math:`f(x, p, a, b, \text{loc}, \text{scale})` is identically equivalent to :math:`\frac{1}{\text{scale}}f(y, p, a, b)` with :math:`y = \frac{1}{\text{scale}}(x - \text{loc})`.

This parameterization derives from the original :math:`(\lambda, \alpha, \beta, \delta, \mu)` parameterization in  Barndorff (1978) by setting:
```

-  :math:`\lambda = p`
-  :math:`\alpha = \frac{a}{\delta} = \frac{\hat{\alpha}}{\delta}`
-  :math:`\beta = \frac{b}{\delta} = \frac{\hat{\beta}}{\delta}`
-  :math:`\delta = \text{scale}`
-  :math:`\mu = \text{location}`

Random variates for the `scipy.stats.genhyperbolic` can be efficiently sampled from the above-mentioned normal variance-mean mixture where `scipy.stats.geninvgauss` is parametrized as :math:`GIG\Big(p = p, b = \sqrt{\hat{\alpha}^2 - \hat{\beta}^2}, \text{loc} = \text{location}, \text{scale} = \frac{1}{\sqrt{\hat{\alpha}^2 - \hat{\beta}^2}}\Big)` so that: :math:`GH(p, \hat{\alpha}, \hat{\beta}) = \hat{\beta} \cdot GIG + \sqrt{GIG} \cdot N(0,1)`

The "generalized" characterization suggests the fact that this distribution is a superclass of several other probability distribution, for instance:

> -  :math:`f(p = \nu,  a = 0, b = 0, \text{loc} = 0, \text{scale} = \sqrt{\delta})` has a Student's t-distribution (`scipy.stats.t`) with :math:`\nu` degrees of freedom.

Reviewer: ```suggestion
-  :math:`f(p = -\nu/2,  a = 0, b = 0, \text{loc} = 0, \text{scale} = \sqrt{\nu})` has a Student's t-distribution (`scipy.stats.t`) with :math:`\nu` degrees of freedom.
```
Please check this.

-  :math:`f(p = 1, a = \hat{\alpha}, b = \hat{\beta}, \text{loc} = \mu, \text{scale} = \delta)` has a Hyperbolic Distribution.
-  :math:`f(p = - 1/2, a = \hat{\alpha}, b = \hat{\beta}, \text{loc} = \mu, \text{scale} = \delta)` has a Normal Inverse Gaussian Distribution (`scipy.stats.norminvgauss`).

> -  :math:`f(p = 1, a = 0, b = 0, loc = \mu, \text{scale} = \delta)` has a Laplace Distribution (`scipy.stats.laplace`) for :math:`\delta \rightarrow 0`

Reviewer: ```suggestion
-  :math:`f(p = 1, a = \delta, b = 0, loc = \mu, \text{scale} = \delta)` has a Laplace Distribution (`scipy.stats.laplace`) for :math:`\delta \rightarrow 0`
```

Author: well spotted thanks!

        {(\sqrt{1 + x^2})^{1/2 - p}}
    \end{eqnarray*}

> The probability density above is defined in the “standardized” form. To shift and/or scale the distribution use the loc and scale parameters. Specifically, :math:`f(x, p, a, b, \text{loc}, \text{scale})` is identically equivalent to :math:`\frac{1}{\text{scale}}f(y, p, a, b)` with :math:`y = \frac{1}{\text{scale}}(x - \text{loc})`.

Reviewer: ```suggestion
The probability density above is defined in the “standardized” form. To shift and/or scale the distribution, use the ``loc`` and ``scale`` parameters. Specifically, :math:`f(x, p, a, b, \texttt{loc}, \texttt{scale})` is identically equivalent to :math:`\frac{1}{\texttt{scale}}f(y, p, a, b)` with :math:`y = \frac{1}{\texttt{scale}}(x - \texttt{loc})`.
```

Reviewer: Meh, never mind.
